### ISSUE TITLE
Provide a way to get extrinsic hash, when submitting a transaction

### ISSUE THREAD

**Issue by mutantcornholio** [polkadot-api/polkadot-api#346]
I'm using code like this, to teleport tokens:
```ts
const tx = api.tx.XcmPallet.limited_teleport_assets(...);
const result = await tx.submit(this.account.address);
```

I would like to have an API to get extrinsic hash for it, which that I can drop into a subscan link, like this one: https://rococo.subscan.io/extrinsic/0x14e2e103b79acc0100612c758721d70862ac6076354ed3d20a29795d3259faab

**Comment by josepot** [comment]
Once #347 gets merged, then it should be possible to create a hash from the extrinsic. Allow me to elaborate: as of today, the only way (using the top-level client API) to get the hash is to do something like this:

```ts
const tx = api.tx.XcmPallet.limited_teleport_assets(...);
const signedTx = await tx.getTx(this.account.address);

const hashedTx = myHashingFunction(signedTx)
```

However, if you do that, then you won't be able to (easily) validate, submit and watch the transaction that you just hashed. Which is, indeed, very unfortunate.

With the changes introduced on #347 you will be able to do either this:

```ts
const tx = api.tx.XcmPallet.limited_teleport_assets(...);
const signedTx = await tx.sign(this.account.address);

const hashedTx = myHashingFunction(signedTx)

const txResult = await client.submit(signedTx)
// of if you like to use observables:
// client.submitAndWatch(signedTx).subscribe(console.log)
```

or this:

```ts
const tx = api.tx.XcmPallet.limited_teleport_assets(...);
tx. signSubmitAndWatch(this.account.address).subscribe(event => {
  if (event.type === "signed") {
    const hashedTx = myHashingFunction(signedTx)
    console.log({hashedTx})
  }
  // Or a strongly-typed switch case over `switch(event.type)`
});
```

Perhaps you think that this still doesn't answer your question, because Polkadot-API is not directly providing the hash that you want.

However, AFAIK there is no such thing as a canonical way of hashing an extrinsic. Please, do correct me if I'm wrong on this, because it's quite possible that I'm mistaken.

What I mean is that the transaction that we provide through these APIs is a blob of SCALE encoded binary data, containing the extrinsic info that it's being broadcasted. You could hash that binary data in a thousand different ways, and to be honest I haven't check yet which one is the hashing method that these block-explorers are using.

The point being that: unless we have some evidence about the existence of a canonical/documented way of hashing the extrinsics, this API is not going to provide a hashed extrinsic (because it can be hashed in a million different ways)... Instead we expose the actual extrinsic and we allow the consumer to hash it however the see fit.

Once again, I could be wrong. Perhaps there is a documented/canonical way of hashing extrinsics. If that's the case then we will make sure that we provide this canonical hashing function from the public API.

**Comment by bkchr** [comment]
> However, AFAIK there is no such thing as a canonical way of hashing an extrinsic. Please, do correct me if I'm wrong on this, because it's quite possible that I'm mistaken.

For sure there is. The hashing used for the extrinsics is the same hashing as configured for the block type. I assume you can extract this from the Metadata and then use it?

**Comment by josepot** [comment]
> The hashing used for the extrinsics is the same hashing as configured for the block type.

I actually have no idea how to get this from the metadata. Could you please help me out here? 🙏 

I mean, in the case of the storage-entires, it's very simple because the `storage.items.type` has a `hashers` property that allows us to know how to properly hash the storage entry.

However, I have no idea how to get the hashing function for the extrinsic from the metadata. I have tried to find this information within the all the different properties of the metadata Extrinsic entry, but I didn't find anything in there...

Then, I thought that perhaps is the same hashing function defined for the storage entry `System.BlockHash`, but that doesn't seem to be the case...

If you could please let me know what's the "proper" way of getting that information from the metadata, that would be greatly appreciated.

**Comment by josepot** [comment]
I mean, I think that I found a rather hacky way to get it:

- First we find the entry in the registry that has a `path` property of: `["frame_system", "EventRecord"]`
- For that entry we look at the type of the parameter named `T`
- That will take us to a lookup entry with a path that looks like this: `["primitive_types", "H256"]`
- Then, we can try to infer the hashing function from the `primitive_type` defined in that `path` property.

It feels a bit hacky, though... Could anyone please let me know if there is a better way of getting the runtime hasher from the metadata? 🙏

**Comment by jsdw** [comment]
I'm not aware of a straightforward way to find out the hasher from the metadata. It's an associated type in some runtime config, and the metadata doesn't contain any information about said types / type names.

In Subxt, the hasher for a chain is configured by the user prior to instantiating a client, ie:

```rust
// We define config for Subxt like this; the goal of the config is to provide 
// the information / types / logic we can't otherwise extract from metadata.
pub enum SubstrateConfig {}
impl subxt::config::Config for SubstrateConfig {
    type Hash = H256;
    type AccountId = AccountId32;
    type Address = MultiAddress<Self::AccountId, u32>;
    type Signature = MultiSignature;
    type Hasher = BlakeTwo256;
    type Header = SubstrateHeader<u32, BlakeTwo256>;
    type ExtrinsicParams = SubstrateExtrinsicParams<Self>;
    type AssetId = u32;
}

// And then later, when we instantiate a client, this config is passed in.
// This config is handed around so that we have access to eg `C::Hasher`
// to know how to hash extrinsics.
let client: OnlineClient::<SubstrateConfig>::from_url("..>").await?;
``` 

I'd very much like to find a way to add this sort of information to the metadata eventually (we were going to look at this sooner but with Tadeo leaving and Alex part-leaving the team, we don't have the bandwidth right now).

If we did add this (maybe in the form of type information for each associated type in each config instance) you could inspect the name of the hasher type in order to know which hashing logic to use. (A more out there idea is that we could provide little WASM bundles or something in or pointed to by the metadata that contain the logic for things like hashing or whatever, so a client could download and use this chain-specific logic without needing to know anything at all up front)

**Comment by josepot** [comment]
As of right now I'm tempted to do something along the lines of this:
- Allow the consumer to pass a hashing function when creating the client.
- If the user has not provided one, then use [the hacky method described in this comment](https://github.com/polkadot-api/polkadot-api/issues/346#issuecomment-2022616069) as an attempt to obtain the hasher from on-chain data.
- If that fails, then we use H256 as the final fallback.

However, I want to sleep on it... We will see.

**Comment by bkchr** [comment]
I think we should just put `Hashing` from the `frame_system::Config` into the metadata. This would make this here much easier.

**Comment by josepot** [comment]
> I think we should just put `Hashing` from the `frame_system::Config` into the metadata. This would make this here much easier.

Works for me!

**Comment by josepot** [comment]
Reopening this, as it was accidentally closed when #347 got merged.

**Comment by josepot** [comment]
Fixed in #507
